src/clj/puppetlabs/services/file_serving/file_metadata/file_metadata_core.clj

```diff
@@ -0,0 +1,200 @@
+            [schema.core :as schema]
+            [cheshire.core :as cheshire]
+            [compojure.core :as compojure :refer [GET ANY PUT]]
+            [liberator.core :refer [defresource]]
+            [liberator.representation :as representation]
+            [liberator.dev :as liberator-dev]
+            [puppetlabs.services.file-serving.config.puppet-fileserver-config-core :as config]
+            [puppetlabs.services.file-serving.utils.posix-utils :as posix-utils]
+            )
+  (:use [clojure.java.io :only (input-stream)]
+        [ring.middleware.params]
+        ))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; 'handler' functions for HTTP endpoints
+
+(defn- checksum-of-file
+  [file]
+  "Compute the md5 checksum of a given file path"
+  (with-open [r (java.nio.file.Files/newInputStream file (into-array java.nio.file.OpenOption []))]
+    {
+     :type "md5"
+     :value (str "{md5}" (DigestUtils/md5Hex r))
+     }
+  ))
+
+(def date-format (SimpleDateFormat. "yyyy.MM.dd HH:mm:ss Z" Locale/ROOT))
+
+(defn date->ruby
+  [time]
+  (.format date-format (Date. time)))
+
+(defn ruby->date
+  [date]
+  (.parse date-format date))
+
+(defn- checksum-of-dir
+  [posix-attr]
+  "Compute the ctime checksum of a given directory"
+    {
+     :type "ctime"
+     :value (str "{ctime}" (date->ruby (.toMillis (posix-attr "creationTime"))))
+     }
+    )
+
+(def base-metadata
+  {
+    :data          {
+                     }
+    :document_type "FileMetadata"
+    :metadata      {
+                     :api_version 1
+                     }
+   })
+
+(defn- get-base-metadata
+  [attrs file-path posix-attrs]
+  (-> attrs
+    (assoc-in [:data :owner] (posix-utils/owner posix-attrs))
+    (assoc-in [:data :group] (posix-utils/group posix-attrs))
+    (assoc-in [:data :mode] (posix-utils/mode posix-attrs))
+    (assoc-in [:data :path] (str (.toAbsolutePath file-path)))))
+
+(defn- get-file-metadata
+  [file-path posix-attrs]
+  (-> (assoc-in base-metadata [:data :checksum] (checksum-of-file file-path))
+      (get-base-metadata file-path posix-attrs)
+      (assoc-in [:data :relative_path] nil)
+      (assoc-in [:data :destination] nil)
+      (assoc-in [:data :type] "file")
+      ))
+
+(defn- get-directory-metadata
+  [file-path posix-attrs]
+  (-> (assoc-in base-metadata [:data :checksum] (checksum-of-dir posix-attrs))
+      (get-base-metadata file-path posix-attrs)
+      (assoc-in [:data :relative_path] ".")
+      (assoc-in [:data :destination] nil)
+      (assoc-in [:data :type] "directory")
+      ))
+
+(defn- get-link-metadata
+  [file-path posix-attrs links]
+  (-> (assoc-in base-metadata [:data :checksum] (checksum-of-file file-path))
+      (get-base-metadata file-path posix-attrs)
+      (assoc-in [:data :relative_path] nil)
+      (assoc-in [:data :destination] (.getCanonicalPath (.toFile (posix-utils/resolve-link file-path))))
+      (assoc-in [:data :type] "link")
+      ))
```

Comment:
Rather than doing `assoc-in` for each of the keys in `:data`, you could use `ks/deep-merge` or do something like.

```clojure
(update-in base-metadata :data #(merge % {:relative_path nil
                                          :destination :nil
                                          :type "file"}))
```